tools/snapshot/snapshot.py

```diff
@@ -0,0 +1,468 @@
+    log.debug(f"successfully cleaned up rclone snapshot credentials on {machine['ip']} (s{machine['shard']})")
+
+
+def cleanup_rclone_config():
+    """
+    Clean/remove rclone credentials that were added (as dictated by the config).
+
+    Note that this will remove the file at whatever the `config_path_on_client`
+    specifies in the config.
+    """
+    threads, pool = [], ThreadPool()
+    for machine in machines:
+        threads.append(pool.apply_async(_cleanup_rclone_config, (machine,)))
+    for t in threads:
+        t.get()
+
+
+def _derive_db_paths(machine):
+    """
+    Internal function to derive the true harmony DB and
+    rsync harmony DB path on the machine.
+    """
+    db_path_on_machine = f"{machine['db_directory']}/harmony_db_{machine['shard']}"
+    db_rsync_path_on_machine = f"{db_path_on_machine}_rsync"
+    return db_path_on_machine, db_rsync_path_on_machine
+
+
+def _bucket_sync(machine):
+    """
+    Internal function to start bucket sync.
+    Function call will block until bucket sync is done on machine.
+
+    Note the convention used when syncing to bucket.
+    """
+    log.debug(f'starting bucket sync on {machine["ip"]} (s{machine["shard"]})')
+    _, rsync_db_path = _derive_db_paths(machine)
+    db_type = 'full' if condition['is_archival'] else 'pruned'
+    bucket, shard = rsync['snapshot_bin'], machine['shard']
+    unix_time, config = int(time.time()), rsync['config_path_on_client']
+    cmd = f"rclone sync {rsync_db_path} {bucket}/{db_type}/{shard}/harmony_db_{shard}.{unix_time} --config {config}"
+    try:
+        _ssh_cmd(machine['user'], machine['ip'], cmd)
+    except subprocess.CalledProcessError as e:
+        log.error("failed to bucket sync db")
+        raise RuntimeError("failed to bucket sync db") from e
+    log.debug(f'successful bucket sync on {machine["ip"]} (s{machine["shard"]})')
+
+
+def _local_sync(machine):
+    """
+    Internal function to trigger a local sink.
+    Function call will block until local sync is done on machine.
+    """
+    log.debug(f'starting local sync on {machine["ip"]} (s{machine["shard"]})')
+    db_path_on_machine, db_rsync_path_on_machine = _derive_db_paths(machine)
+    cmd = f"rclone sync {db_path_on_machine} {db_rsync_path_on_machine} --transfers 64"
+    try:
+        _ssh_cmd(machine['user'], machine['ip'], cmd)
+    except subprocess.CalledProcessError as e:
+        log.error("failed to local sync db")
+        raise RuntimeError("failed to local sync db") from e
+    log.debug(f'successful local sync on {machine["ip"]} (s{machine["shard"]})')
+
+
+def _stop_harmony(machine):
+    """
+    Internal function to stop and verify harmony service.
+    Assumption is that harmony is ran as a service.
+    """
+    log.debug(f'stopping harmony service on {machine["ip"]} (s{machine["shard"]})')
+    machine_stop_response = _ssh_cmd(machine['user'], machine['ip'], "sudo systemctl stop harmony")
+    time.sleep(1)  # Wait for graceful shutdown
```

Comment:
may need to wait a few loop before gracefully shutdown though. in most of cases, 1s should be enough though.

Reply:
I will add a timeout logic here. So short circuit return if the process stopped, otherwise keep checking for up to 5 seconds (at which point, if not shut down, will raise an exception to trigger node reset to working state)